ruby_programming/automated_testing/test_driven_development.md

```diff
@@ -0,0 +1,206 @@
+
+This development cycle is known as red-green-refactor, and it’s at the heart of the TDD process. When following TDD you will iterate over each unit of work needed for your project with these three steps until you’re done!
+
+One key aspect of the red-green-refactor cycle that *isn’t* in the name, is that the code you write to go from `red` to `green` should be the **minimum** amount required to pass the test. If you find that the functionality you’re adding actually does *more* than is being tested, that is a sign that your method is likely doing too much, or possibly that your tests aren’t testing all of the right functionality.
+
+#### TDD Examples
+Let’s walk through a TDD approach to a basic class.
+
+Given the general problem:
+
+Create a `Square` class that takes an integer on initialization for the length of each side, and has an `#area` method and a `#perimeter` method which returns (as you probably guessed) the area and perimeter of the square respectively.
+
+Start by making a directory to house this example, and then `lib` and `spec` directories to house the class and the specs respectively. Next, create the spec and class files. Lastly, open the example directory in your text editor so that you can start this TDD style!
+
+~~~ bash
+mkdir tdd-lesson
+cd tdd-lesson
+mkdir lib
+mkdir spec
+touch lib/square.rb
+touch spec/square_spec.rb
+code .
+~~~
+
+Now, how might you want to TDD the creation of this class? Well, you know that you need to `describe` a class called `Square`, and that you'll need to describe its `#area` method, so start with setting all that up in a `square_spec.rb` file:
+
+~~~rb
+require_relative '../lib/square'
+
+describe Square do
+  describe "#area" do
+
+  end
+end
+~~~
+
+Since `Square` will be initialized with a side length, do that initialization in a couple of different `context` blocks so that you can initialize differently sized squares to test the method:
+
+~~~rb
+require_relative '../lib/square'
+
+describe Square do
+  describe "#area" do
+    context 'when side length is 4' do
+      let(:size_4_square) { described_class.new(4) }
+    end
+
+    context 'when side length is 6' do
+      let(:size_6_square) { described_class.new(6) }
+    end
+  end
+end
+~~~
+
+Next add an `it` block to each of the contexts to actually test that the method works as you expect:
+
+~~~rb
+require_relative '../lib/square'
+
+describe Square do
+  describe "#area" do
+    context 'when side length is 4' do
+      let(:size_4_square) { described_class.new(4) }
+
+      it ‘returns 16’ do
+        expect(size_4_square.area).to eq(16)
+      end
+    end
+
+    context 'when side length is 6' do
+      let(:size_6_square) { described_class.new(6) }
+
+      it ‘returns 36’ do
+        expect(size_6_square.area).to eq(36)
+      end
+    end
+  end
+end
+~~~
+
+Alright, that should have you set up to test `#area`, now run the tests by entering `rspec` in the terminal!
+
```

Comment:
I think it would be nice to show the expected test output, so that they get used to interpreting the error messages. This might even be good to add after each time you have them run `rspec`